File: src/services/charging-service.ts
/**
 * Charging Service - Simple API for External Services
 *
 * This is the main interface for payment services and other parts
 * of the application to control charging sessions without knowing
 * OCPP protocol details.
 */

import { database } from '../database/mock-database';
import { connectionManager } from '../ocpp/connection-manager';
import {
  ChargePointNotFoundError,
  ChargePointOfflineError,
  ConnectorNotFoundError,
  ConnectorUnavailableError,
  SessionNotFoundError,
  SessionAlreadyActiveError,
  RemoteOperationFailedError,
  OperationTimeoutError,
  ChargePointNotConfiguredError
} from '../errors/custom-errors';
import {
  TurnOnResult,
  TurnOffResult,
  EnergyConsumptionResult
} from '../types/domain-types';
import { RemoteStartStopStatus } from '../types/ocpp-types';

export class ChargingService {
  /**
   * Turn on a charging socket (start a charging session)
   *
   * @param chargePointId - ID of the charge point
   * @param connectorId - Connector number (1, 2, etc.)
   * @param userId - User/customer ID who is starting the session
   * @returns Promise with transaction details
   * @throws ChargePointNotFoundError - If charge point doesn't exist
   * @throws ChargePointOfflineError - If charge point is not connected
   * @throws ConnectorNotFoundError - If connector doesn't exist
   * @throws ConnectorUnavailableError - If connector is busy/faulted
   * @throws SessionAlreadyActiveError - If connector already has active session
   * @throws RemoteOperationFailedError - If charge point rejects the command
   * @throws OperationTimeoutError - If operation times out
   */
  async turnOn(
    chargePointId: string,
    connectorId: number,
    userId: string
  ): Promise<TurnOnResult> {
    // 1. Check if charge point exists in database
    const chargePoint = database.getChargePoint(chargePointId);
    if (!chargePoint) {
      throw new ChargePointNotFoundError(chargePointId);
    }

    // 2. Check if charge point is configured (not pending)
    if (chargePoint.registrationStatus !== 'accepted') {
      throw new ChargePointNotConfiguredError(chargePointId);
    }

    // 3. Check if charge point is connected
    if (!connectionManager.isConnected(chargePointId)) {
      throw new ChargePointOfflineError(chargePointId);
    }

    // 4. Check if connector exists
    const connector = database.getConnector(chargePointId, connectorId);
    if (!connector) {
      throw new ConnectorNotFoundError(chargePointId, connectorId);
    }

    // 5. Check connector status
    if (connector.status === 'Faulted') {
      throw new ConnectorUnavailableError(chargePointId, connectorId, 'Faulted');
    }

    if (connector.status === 'Unavailable') {
      throw new ConnectorUnavailableError(chargePointId, connectorId, 'Unavailable');
    }

    // 6. Check if connector already has an active session
    const existingSession = database.getActiveSessionByConnector(chargePointId, connectorId);
    if (existingSession) {
      throw new SessionAlreadyActiveError(chargePointId, connectorId);
    }

    // 7. Send RemoteStartTransaction to charge point
    try {
      const response = await connectionManager.remoteStartTransaction(
        chargePointId,
        userId,
        connectorId
      );

      if (response.status === RemoteStartStopStatus.Rejected) {
        throw new RemoteOperationFailedError('start', chargePointId, 'Rejected by charge point');
      }

      // 8. Wait a bit for the StartTransaction message from the charge point
      await this.waitForSessionStart(chargePointId, connectorId, 10000);

      // 9. Get the created transaction
      const session = database.getActiveSessionByConnector(chargePointId, connectorId);
      if (!session) {
        throw new RemoteOperationFailedError(
          'start',
          chargePointId,
          'Session not created after remote start'
        );
      }

      return {
        success: true,
        transactionId: session.transactionId,
        chargePointId,
        connectorId,
        userId,
        message: 'Charging session started successfully'
      };
    } catch (error) {
      if (error instanceof Error && error.message.includes('timeout')) {
        throw new OperationTimeoutError('RemoteStartTransaction', chargePointId);
      }
      throw error;
    }
  }

  /**
   * Turn off a charging socket (stop a charging session)
   *
   * @param chargePointId - ID of the charge point
   * @param transactionId - Transaction ID to stop
   * @returns Promise with energy consumption details
   * @throws ChargePointNotFoundError - If charge point doesn't exist
   * @throws ChargePointOfflineError - If charge point is not connected
   * @throws SessionNotFoundError - If transaction doesn't exist
   * @throws RemoteOperationFailedError - If charge point rejects the command
   * @throws OperationTimeoutError - If operation times out
   */
  async turnOff(chargePointId: string, transactionId: number): Promise<TurnOffResult> {
    // 1. Check if session exists
    const session = database.getSession(transactionId);
    if (!session) {
      throw new SessionNotFoundError(transactionId);
    }

    // Verify it's the right charge point
    if (session.chargePointId !== chargePointId) {
      throw new SessionNotFoundError(transactionId);
    }

    // 2. Check if session is still active
    if (session.status !== 'active') {
      // Session already completed, return the final data
      const energyConsumed = (session.endMeterValue || session.currentMeterValue) - session.startMeterValue;
      const duration = session.endTime
        ? (session.endTime.getTime() - session.startTime.getTime()) / 1000
        : 0;

      return {
        success: true,
        transactionId,
        chargePointId,
        energyConsumed,
        duration,
        message: 'Session was already stopped'
      };
    }

    // 3. Check if charge point is connected
    if (!connectionManager.isConnected(chargePointId)) {
      throw new ChargePointOfflineError(chargePointId);
    }

    // 4. Send RemoteStopTransaction to charge point
    try {
      const response = await connectionManager.remoteStopTransaction(chargePointId, transactionId);

      if (response.status === RemoteStartStopStatus.Rejected) {
        throw new RemoteOperationFailedError('stop', chargePointId, 'Rejected by charge point');
      }

      // 5. Wait for the StopTransaction message from the charge point
      await this.waitForSessionStop(transactionId, 15000);

      // 6. Get final session data
      const updatedSession = database.getSession(transactionId);
      if (!updatedSession) {
        throw new SessionNotFoundError(transactionId);
      }

      const energyConsumed =
        (updatedSession.endMeterValue || updatedSession.currentMeterValue) -
        updatedSession.startMeterValue;

      const duration = updatedSession.endTime
        ? (updatedSession.endTime.getTime() - updatedSession.startTime.getTime()) / 1000
        : (new Date().getTime() - updatedSession.startTime.getTime()) / 1000;

      return {
        success: true,
        transactionId,
        chargePointId,
        energyConsumed,
        duration,
        message: 'Charging session stopped successfully'
      };
    } catch (error) {
      if (error instanceof Error && error.message.includes('timeout')) {
        throw new OperationTimeoutError('RemoteStopTransaction', chargePointId);
      }
      throw error;
    }
  }

  /**
   * Get current energy consumption for an active charging session
   *
   * @param chargePointId - ID of the charge point
   * @param transactionId - Transaction ID
   * @returns Promise with energy consumption data
   * @throws SessionNotFoundError - If transaction doesn't exist
   */
  async getEnergyConsumption(
    chargePointId: string,
    transactionId: number
  ): Promise<EnergyConsumptionResult> {
    // 1. Get session
    const session = database.getSession(transactionId);
    if (!session) {
      throw new SessionNotFoundError(transactionId);
    }

    // Verify it's the right charge point
    if (session.chargePointId !== chargePointId) {
      throw new SessionNotFoundError(transactionId);
    }

    // 2. Calculate current energy consumption
    const currentMeterValue =
      session.status === 'completed' && session.endMeterValue
        ? session.endMeterValue
        : session.currentMeterValue;

    const energyConsumed = currentMeterValue - session.startMeterValue;

    const endTime = session.status === 'completed' && session.endTime ? session.endTime : new Date();
    const durationSeconds = Math.floor((endTime.getTime() - session.startTime.getTime()) / 1000);

    // 3. Get latest energy reading for additional details
    const latestReading = database.getLatestEnergyReading(transactionId);

    return {
      transactionId,
      chargePointId: session.chargePointId,
      connectorId: session.connectorId,
      userId: session.userId,
      startTime: session.startTime,
      currentEnergyWh: energyConsumed,
      durationSeconds,
      status: session.status === 'completed' ? 'completed' : 'active',
      lastUpdate: latestReading?.timestamp || session.startTime
    };
  }

  /**
   * Helper: Wait for session to start
   */
  private async waitForSessionStart(
    chargePointId: string,
    connectorId: number,
    timeoutMs: number
  ): Promise<void> {
    const startTime = Date.now();
    const pollInterval = 500; // Check every 500ms

    while (Date.now() - startTime < timeoutMs) {
      const session = database.getActiveSessionByConnector(chargePointId, connectorId);
      if (session) {
        return;
      }
      await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }

    // Timeout
    throw new Error('Timeout waiting for session to start');
  }

  /**
   * Helper: Wait for session to stop
   */
  private async waitForSessionStop(transactionId: number, timeoutMs: number): Promise<void> {
    const startTime = Date.now();
    const pollInterval = 500; // Check every 500ms

    while (Date.now() - startTime < timeoutMs) {
      const session = database.getSession(transactionId);
      if (session && session.status === 'completed') {
        return;
      }
      await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }

    // Timeout - session might still be stopping
    // We'll just return what we have
  }

  /**
   * Get all active sessions
   */
  async getActiveSessions(): Promise<EnergyConsumptionResult[]> {
    const sessions = database.getAllActiveSessions();
    return sessions.map((session) => {
      const energyConsumed = session.currentMeterValue - session.startMeterValue;
      const durationSeconds = Math.floor((Date.now() - session.startTime.getTime()) / 1000);
      const latestReading = database.getLatestEnergyReading(session.transactionId);

      return {
        transactionId: session.transactionId,
        chargePointId: session.chargePointId,
        connectorId: session.connectorId,
        userId: session.userId,
        startTime: session.startTime,
        currentEnergyWh: energyConsumed,
        durationSeconds,
        status: 'active',
        lastUpdate: latestReading?.timestamp || session.startTime
      };
    });
  }
}

export const chargingService = new ChargingService();

File: pkg/service/charging_service.go
package service

import (
	"time"

	"github.com/ADEXITUM/ocpp_management_system/pkg/database"
	"github.com/ADEXITUM/ocpp_management_system/pkg/errors"
	"github.com/ADEXITUM/ocpp_management_system/pkg/ocpp"
	"github.com/ADEXITUM/ocpp_management_system/pkg/types"
)

// ChargingService provides simple API for controlling charging sessions
type ChargingService struct {
	db                *database.MockDatabase
	connectionManager *ocpp.ConnectionManager
}

// NewChargingService creates a new charging service
func NewChargingService(db *database.MockDatabase, cm *ocpp.ConnectionManager) *ChargingService {
	return &ChargingService{
		db:                db,
		connectionManager: cm,
	}
}

// TurnOn starts a charging session
func (s *ChargingService) TurnOn(
	chargePointID string,
	connectorID int,
	userID string,
) (*types.TurnOnResult, error) {
	// 1. Check if charge point exists in database
	chargePoint := s.db.GetChargePoint(chargePointID)
	if chargePoint == nil {
		return nil, errors.NewChargePointNotFoundError(chargePointID)
	}

	// 2. Check if charge point is configured
	if chargePoint.RegistrationStatus != "accepted" {
		return nil, errors.NewChargePointNotConfiguredError(chargePointID)
	}

	// 3. Check if charge point is connected
	if !s.connectionManager.IsConnected(chargePointID) {
		return nil, errors.NewChargePointOfflineError(chargePointID)
	}

	// 4. Check if connector exists
	connector := s.db.GetConnector(chargePointID, connectorID)
	if connector == nil {
		return nil, errors.NewConnectorNotFoundError(chargePointID, connectorID)
	}

	// 5. Check connector status
	if connector.Status == types.StatusFaulted {
		return nil, errors.NewConnectorUnavailableError(chargePointID, connectorID, "Faulted")
	}

	if connector.Status == types.StatusUnavailable {
		return nil, errors.NewConnectorUnavailableError(chargePointID, connectorID, "Unavailable")
	}

	// 6. Check if connector already has an active session
	existingSession := s.db.GetActiveSessionByConnector(chargePointID, connectorID)
	if existingSession != nil {
		return nil, errors.NewSessionAlreadyActiveError(chargePointID, connectorID)
	}

	// 7. Send RemoteStartTransaction to charge point
	response, err := s.connectionManager.RemoteStartTransaction(
		chargePointID,
		userID,
		&connectorID,
	)

	if err != nil {
		if isTimeoutError(err) {
			return nil, errors.NewOperationTimeoutError("RemoteStartTransaction", chargePointID)
		}
		return nil, err
	}

	if response.Status == types.RemoteStartStopRejected {
		return nil, errors.NewRemoteOperationFailedError("start", chargePointID, "Rejected by charge point")
	}

	// 8. Wait for the StartTransaction message from the charge point.
	// Real devices/testers can take longer between RemoteStart acceptance and StartTransaction.
	session, err := s.waitForSessionStart(chargePointID, connectorID, 90*time.Second)
	if err != nil {
		return nil, err
	}

	return &types.TurnOnResult{
		Success:       true,
		TransactionID: session.TransactionID,
		ChargePointID: chargePointID,
		ConnectorID:   connectorID,
		UserID:        userID,
		Message:       "Charging session started successfully",
	}, nil
}

// TurnOff stops a charging session
func (s *ChargingService) TurnOff(
	chargePointID string,
	transactionID int,
) (*types.TurnOffResult, error) {
	// 1. Check if session exists
	session := s.db.GetSession(transactionID)
	if session == nil {
		return nil, errors.NewSessionNotFoundError(transactionID)
	}

	// Verify it's the right charge point
	if session.ChargePointID != chargePointID {
		return nil, errors.NewSessionNotFoundError(transactionID)
	}

	// 2. Check if session is still active
	if session.Status != "active" {
		// Session already completed, return the final data
		energyConsumed := float64(getEndMeterValue(session) - session.StartMeterValue)
		duration := getDuration(session)

		return &types.TurnOffResult{
			Success:        true,
			TransactionID:  transactionID,
			ChargePointID:  chargePointID,
			EnergyConsumed: energyConsumed,
			Duration:       duration,
			Message:        "Session was already stopped",
		}, nil
	}

	// 3. Check if charge point is connected
	if !s.connectionManager.IsConnected(chargePointID) {
		return nil, errors.NewChargePointOfflineError(chargePointID)
	}

	// 4. Send RemoteStopTransaction to charge point
	response, err := s.connectionManager.RemoteStopTransaction(chargePointID, transactionID)
	if err != nil {
		if isTimeoutError(err) {
			return nil, errors.NewOperationTimeoutError("RemoteStopTransaction", chargePointID)
		}
		return nil, err
	}

	if response.Status == types.RemoteStartStopRejected {
		return nil, errors.NewRemoteOperationFailedError("stop", chargePointID, "Rejected by charge point")
	}

	// 5. Wait for the StopTransaction message from the charge point
	s.waitForSessionStop(transactionID, 15*time.Second)

	// 6. Get final session data
	updatedSession := s.db.GetSession(transactionID)
	if updatedSession == nil {
		return nil, errors.NewSessionNotFoundError(transactionID)
	}

	energyConsumed := float64(getEndMeterValue(updatedSession) - updatedSession.StartMeterValue)
	duration := getDuration(updatedSession)

	return &types.TurnOffResult{
		Success:        true,
		TransactionID:  transactionID,
		ChargePointID:  chargePointID,
		EnergyConsumed: energyConsumed,
		Duration:       duration,
		Message:        "Charging session stopped successfully",
	}, nil
}

// GetEnergyConsumption gets current energy consumption for a session
func (s *ChargingService) GetEnergyConsumption(
	chargePointID string,
	transactionID int,
) (*types.EnergyConsumptionResult, error) {
	// 1. Get session
	session := s.db.GetSession(transactionID)
	if session == nil {
		return nil, errors.NewSessionNotFoundError(transactionID)
	}

	// Verify it's the right charge point
	if session.ChargePointID != chargePointID {
		return nil, errors.NewSessionNotFoundError(transactionID)
	}

	// 2. Calculate current energy consumption
	currentMeterValue := session.CurrentMeterValue
	if session.Status == "completed" && session.EndMeterValue != nil {
		currentMeterValue = *session.EndMeterValue
	}

	energyConsumed := float64(currentMeterValue - session.StartMeterValue)
	durationSeconds := getDuration(session)

	// 3. Get latest energy reading for additional details
	latestReading := s.db.GetLatestEnergyReading(transactionID)
	lastUpdate := session.StartTime
	if latestReading != nil {
		lastUpdate = latestReading.Timestamp
	}

	status := "active"
	if session.Status == "completed" {
		status = "completed"
	}

	return &types.EnergyConsumptionResult{
		TransactionID:   transactionID,
		ChargePointID:   session.ChargePointID,
		ConnectorID:     session.ConnectorID,
		UserID:          session.UserID,
		StartTime:       session.StartTime,
		CurrentEnergyWh: energyConsumed,
		DurationSeconds: durationSeconds,
		Status:          status,
		LastUpdate:      lastUpdate,
	}, nil
}

// GetActiveSessions returns all active charging sessions
func (s *ChargingService) GetActiveSessions() []*types.EnergyConsumptionResult {
	sessions := s.db.GetAllActiveSessions()
	results := make([]*types.EnergyConsumptionResult, 0, len(sessions))

	for _, session := range sessions {
		energyConsumed := float64(session.CurrentMeterValue - session.StartMeterValue)
		durationSeconds := int(time.Since(session.StartTime).Seconds())

		latestReading := s.db.GetLatestEnergyReading(session.TransactionID)
		lastUpdate := session.StartTime
		if latestReading != nil {
			lastUpdate = latestReading.Timestamp
		}

		results = append(results, &types.EnergyConsumptionResult{
			TransactionID:   session.TransactionID,
			ChargePointID:   session.ChargePointID,
			ConnectorID:     session.ConnectorID,
			UserID:          session.UserID,
			StartTime:       session.StartTime,
			CurrentEnergyWh: energyConsumed,
			DurationSeconds: durationSeconds,
			Status:          "active",
			LastUpdate:      lastUpdate,
		})
	}

	return results
}

// Helper functions

func (s *ChargingService) waitForSessionStart(
	chargePointID string,
	connectorID int,
	timeout time.Duration,
) (*types.ChargingSession, error) {
	deadline := time.Now().Add(timeout)
	pollInterval := 500 * time.Millisecond

	for time.Now().Before(deadline) {
		session := s.db.GetActiveSessionByConnector(chargePointID, connectorID)
		if session != nil {
			return session, nil
		}
		time.Sleep(pollInterval)
	}

	return nil, errors.NewRemoteOperationFailedError(
		"start",
		chargePointID,
		"Session not created after remote start",
	)
}

func (s *ChargingService) waitForSessionStop(transactionID int, timeout time.Duration) {
	deadline := time.Now().Add(timeout)
	pollInterval := 500 * time.Millisecond

	for time.Now().Before(deadline) {
		session := s.db.GetSession(transactionID)
		if session != nil && session.Status == "completed" {
			return
		}
		time.Sleep(pollInterval)
	}
}

func getEndMeterValue(session *types.ChargingSession) int {
	if session.EndMeterValue != nil {
		return *session.EndMeterValue
	}
	return session.CurrentMeterValue
}

func getDuration(session *types.ChargingSession) int {
	if session.EndTime != nil {
		return int(session.EndTime.Sub(session.StartTime).Seconds())
	}
	return int(time.Since(session.StartTime).Seconds())
}

func isTimeoutError(err error) bool {
	if err == nil {
		return false
	}
	return err.Error() == "request timeout" ||
		(len(err.Error()) > 15 && err.Error()[:15] == "request timeout")
}
